recon.py

```diff
@@ -0,0 +1,329 @@
+"""
+A Proposal for reconciliation graph and reconciliation interface.
+"""
+from abc import ABC  # Abstract Base Classes
+from typing import List, Dict
+from enum import Enum
+
+__all__ = ['MappingNode', 'EventType', 'Event', 'Cospeciation', 'Duplication', 'Transfer', 'Loss',
+           'TipTip', 'Reconciliation', 'ReconGraph']
+
+class MappingNode:
+    """
+    MappingNode is a node in the reconciliation graph that maps
+    a parasite to a host.
+    """
+    def __init__(self, parasite_vertex: str, host_vertex: str):
+        self._parasite = parasite_vertex
+        self._host = host_vertex
+
+    @property
+    def parasite(self) -> str:
+        return self._parasite
+
+    @property
+    def host(self) -> str:
+        return self._host
+    
+    def __eq__(self, other):
+        if type(self) == type(other):
+            return self._parasite == other.parasite and self._host == other.host
+        return False
+    
+    def __hash__(self):
+        return hash((self._parasite, self._host))
+    
+    def __repr__(self):
+        return "%s(%s, %s)" % (type(self).__name__, self._parasite, self._host)
+
+class EventType(Enum):
+    """
+    EventType is the type of event node.
+    """
+    COSPECIATION = 1
+    DUPLICATION = 2
+    TRANSFER = 3
+    LOSS = 4
+    TIPTIP = 5
+
+class Event(ABC):
+    """
+    Base Class for events.
+    """
+    def __init__(self, freq: float = None):
+        self._freq = freq
+
+    @property
+    def freq(self) -> float:
+        return self._freq
+
+    @property
+    def event_type(self) -> EventType:
+        """
+        Returns the event type of this event.
+        """
+        raise NotImplementedError("Event is an abstract base class")
+
+class TwoChildrenEvent(Event, ABC):
+    """
+    Base Class for events with two children.
+    """
+    def __init__(self, left: MappingNode, right: MappingNode, freq=None):
+        """
+        Creates a Cospeciation, Duplication, or Transfer event. The event node
+        has two children MappingNodes called left and right.
+        """
+        super().__init__(freq)
+        self._left: MappingNode = left
+        self._right: MappingNode = right
+
+    @property
+    def left(self) -> MappingNode:
+        return self._left
+
+    @property
+    def right(self) -> MappingNode:
+        return self._right
+    
+    def __eq__(self, other):
+        if type(self) == type(other):
+            return self._left == other.left and self._right == other.right
+        return False
+    
+    def __hash__(self):
+        return hash((self._left, self._right))
+    
+    def __repr__(self):
+        return "%s(%s, %s)" % (type(self).__name__, self._left, self._right)
+
+class Cospeciation(TwoChildrenEvent):
```

Comment:
Currently, the documentation does not automatically say which class this class inherits from.

Reply:
We might have to modify the docs directly.